rclcpp/src/rclcpp/callback_group.cpp

```diff
@@ -97,6 +95,22 @@ CallbackGroup::automatically_add_to_executor_with_node() const
   return automatically_add_to_executor_with_node_;
 }
 
+std::shared_ptr<rclcpp::GuardCondition>
+CallbackGroup::create_notify_guard_condition(
+  const rclcpp::Context::SharedPtr context_ptr)
+{
+  notify_guard_condition_ =
+    std::make_unique<rclcpp::GuardCondition>(context_ptr);
```

Comment:
I don't love that this implicitly resets the notify guard condition if it was already set. It sounds like a subtle bug waiting to happen.

I kind of think that we should either have an interface like this:

- `create_notify_guard_condition()` -> throws if already has one
- `destroy_notify_guard_condition()` -> throws if not already set
- `get_notify_guard_condition()` -> could return nullptr

Or something like this:

- `exchange_notify_guard_condition()` -> returns previous pointer value, could be nullptr
  - the guard condition would be created by the executor and passed in here
- `get_notify_guard_condition()` -> can return nullptr
  - this could store/return a `weak_ptr`

Both of these are more explicit about changing controlling the lifecycle of the shared pointer.

Reply:
The function is not intended to create the guard condition multiple times in the callback group. I'll update from `create_notify_guard_condition` to `get_notify_guard_condition`, and also add a check statement.

Reply:
Sorry, I didn't think over your suggestions before.
I think the first one you suggested is good. Thank you for the detailed guidance.